### ISSUE TITLE
Create small Web Worker harness for invoking WASM transcoder

### ISSUE THREAD

**Issue by kenrussell** [BinomialLLC/basis_universal#21]
With the nice contribution in #7 , the WASM version of the Basis Universal transcoder is now practical to deploy along with web apps and libraries. Performance measurements (thanks @austinEng) show that it's within about 50% of the performance of the native version if compiled directly into the Chrome browser.

It would be beneficial to the community if a small harness could be written which would instantiate a web worker, load the WASM module on it, and define an API on the calling thread (and associated messages back and forth to the worker) so that the transcoder can easily be invoked without blocking the main thread.

Another thing to consider is whether fetch functionality should be incorporated, so that the worker can both fetch a Basis file and transcode slices out of it on demand, without bouncing all of the Basis file's data back to the main thread. (Only the transcoded data would be, and it would be transferred.)

CC @shrekshao @jdashg @RafaelCintron @grorg @donmccurdy .

**Comment by jdashg** [comment]
Cool!

**Comment by donmccurdy** [comment]
We've added a [THREE.BasisTextureLoader](https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/BasisTextureLoader.js) wrapper in the three.js library with some of these enhancements. The API is:

```js
const loader = new THREE.BasisTextureLoader();

// Detecting supported formats requires a WebGL context.
// The transcoder path should contain a JS wrapper and
// WASM binary with predetermined names. The maxWorkers
// option is used to avoid conflicts with other application
// services that may need workers.
loader.detectSupport( renderer );
loader.setTranscoderPath( './basis/' );
loader.setMaxWorkers( navigator.hardwareConcurrency || 4 );

loader.load( './assets/diffuse.basis', ( texture ) => {

	material.map = texture;
	
	render();

} );
```

There's still a good bit of work left to do on that, including support for mipmaps, allowing developers to choose the output format, and allowing a single loader to transcode to different output formats in parallel, e.g. when parsing different textures from a single model. (https://github.com/mrdoob/three.js/issues/16524)

As I assume you're suggesting, this could certainly be implemented in a library-agnostic harness instead. It would be helpful to know how much interest there is in a standalone implementation – would BabylonJS use it, or write their own worker management? If three.js devs will be the primary consumers and maintainers of the harness, it would be easier for us to write and maintain the code in the three.js repository instead, but I don't want to block this idea either.

When wider browser support for WASM multithreading arrives, that may be more efficient and cleaner to integrate than using Web Workers.

> Another thing to consider is whether fetch functionality should be incorporated...

So far I haven't seen a strong case that fetching an ArrayBuffer on a worker improves outcomes. The `fetch()` API is non-blocking, aside from maybe a fraction of a millisecond to begin the request. If the resource can be fetched on the main thread without blocking, then transferred to the worker without copying, the additional complexity of fetching from the worker is probably not needed. https://stackoverflow.com/a/45392971/1314762

**Comment by kenrussell** [comment]
@donmccurdy I expect there would be interest in a standalone, worker-based transcoder from many other engines like Babylon.js, Pixi.js, PlayCanvas, and more. If the API were fairly low-level and supported opening a Basis file by transferring in an ArrayBuffer, transcoding individual slices (receiving a Promise of an ArrayBuffer back), and closing the file, I'd hope that Three's BasisLoader could be implemented on top of it fairly easily. This API would implicitly support mipmaps, since each would be transcoded separately by the application.

Thanks for sharing your experience with Fetch and workers. Given that, I agree that the API should be kept simple.

Any contribution you'd be willing to make (based on Three's BasisLoader) would be greatly appreciated.

**Comment by shalzz** [comment]
@kenrussell Can you please share the performance measurements comparing the webassembly transcoder with a native version if compiled directly into the Chrome browser?

**Comment by kenrussell** [comment]
@shalzz as mentioned above, the WebAssembly version of the Basis Universal transcoder is within 50% of the performance of compiling it in to the browser. The browser vendors discussed this and decided it wasn't worth the complexity to expose Basis Universal as a compressed texture format directly from the browser.

**Comment by shalzz** [comment]
@kenrussell Okay. I'm just asking for the source/references :)

**Comment by austinEng** [comment]
Hey @shalzz, I did the work on prototyping Chromium integration. I tested transcoding times for various texture dimensions and texture formats. The results shared in the WebGL working group are [here](https://docs.google.com/spreadsheets/d/1ZAOF36UEr1Jmlm8ZeK9SNuZKcTSm-otLN4aqHc1OSeI/edit?usp=sharing). The native implementation was 1x to 2x faster than the WASM implementation.

**Comment by toji** [comment]
Sounds like this issue should be resolved now that #127 has been merged.

**Comment by kenrussell** [comment]
Yes! Thanks Brandon for that nice contribution!
